go/pools/resource_pool.go

```diff
@@ -142,9 +156,40 @@ func NewResourcePool(factory Factory, capacity, maxCap int, idleTimeout time.Dur
 		rp.idleTimer = timer.NewTimer(idleTimeout / 10)
 		rp.idleTimer.Start(rp.closeIdleResources)
 	}
+
+	if refreshCheck != nil && refreshInterval > 0 {
+		rp.refreshInterval = refreshInterval
+		rp.refreshCheck = refreshCheck
+		rp.startRefreshTicker()
+	}
+
 	return rp
 }
 
+func (rp *ResourcePool) startRefreshTicker() {
+	rp.refreshTicker = time.NewTicker(rp.refreshInterval)
+	rp.refreshStop = make(chan struct{})
+	rp.refreshWg.Add(1)
+	go func() {
+		defer rp.refreshWg.Done()
+		for {
+			select {
+			case <-rp.refreshTicker.C:
+				val, err := rp.refreshCheck()
+				if err != nil {
+					log.Info(err)
+				}
+				if val {
+					go rp.reopen()
```

Comment:
I don't see anything stopping two goroutines both running `rp.reopen()` at the same time. one could start the reopen, and later a second attempt might make it in before the first one is finished

should we protect against this with a mutex or something?

Reply:
Yeah, it's possible in theory;  but since reopen() itself shuts down the refresh ticker almost immediately, it's probably highly unlikely to get executed again until reopen restarts the ticker and exits.  I'll think about it, I guess a local mutex in this closure will do the trick.